src/main/java/org/sagebionetworks/bridge/notification/helper/TemplateVariableHelper.java

```diff
@@ -0,0 +1,118 @@
+        // participant, it might appear that the survey is scheduled to the day _before_ the user is created. To
+        // account for this, the scheduledOnStartTime should be the day before the user is created.
+        Iterator<ScheduledActivity> surveyIterator = bridgeHelper.getSurveyHistory(studyId, userId,
+                workerConfig.getEngagementSurveyGuid(), participantCreatedOn.minusDays(1),
+                participantCreatedOn.plusDays(30));
+
+        // Engagement survey is scheduled upon enrollment, and is scheduled exactly once. There should be exactly one
+        // engagement survey in the results.
+        if (!surveyIterator.hasNext()) {
+            throw new IllegalStateException("User " + userId + " does not have an engagement survey");
+        }
+        ScheduledActivity engagementSurveyActivity = surveyIterator.next();
+        String studyCommitment = getStudyCommitmentFromActivity(userId, engagementSurveyActivity,
+                KEY_BENEFITS);
+
+        // Replace value.
+        return message.replace(TEMPLATE_VAR_STUDY_COMMITMENT, studyCommitment);
+    }
+
+    // Helper method that, given a ScheduledActivity and a list of keys, extracts the ClientData and recursively
+    // follows keys to extract the expected result. Throws if the key or value are missing from the ClientData. This
+    // method exists primarily to reduce code duplication and make it easier to modify if needed.
+    @SuppressWarnings("unchecked")
+    private String getStudyCommitmentFromActivity(String userId, ScheduledActivity activity, String... keys) {
+        Object obj = activity.getClientData();
+        if (obj == null) {
+            throw new IllegalStateException("User " + userId + " has no client data for engagement survey");
+        }
+        obj = RestUtils.toType(obj, Map.class);
+
+        // Recursive descent on the map.
+        for (String oneKey : keys) {
+            Map<String, Object> map = (Map<String, Object>) obj;
+            if (map.isEmpty()) {
+                throw new IllegalStateException("User " + userId + " engagement survey client data has no key " +
+                        oneKey);
+            }
+            obj = map.get(oneKey);
+            if (obj == null) {
+                throw new IllegalStateException("User " + userId + " engagement survey client data has no value for " +
+                        "key " + oneKey);
+            }
+        }
+
+        return String.valueOf(obj);
+    }
+
+    // Helper method that resolve ${url}.
+    private String resolveUrlVariable(String studyId, String message) {
+        // Short-cut: No template variable to resolve.
+        if (!message.contains(TEMPLATE_VAR_URL)) {
+            return message;
+        }
+
+        // Replace value.
+        WorkerConfig workerConfig = dynamoHelper.getNotificationConfigForStudy(studyId);
+        return message.replace(TEMPLATE_VAR_URL, workerConfig.getAppUrl());
```

Comment:
Any chance it could be in string more than once? replaceAll()?

Reply:
According to javadoc, replace() replaces all instances of the string. Updated test to verify that this is the case.